src/lua/datetime.lua

```diff
@@ -0,0 +1,623 @@
+        secs = secs + 1
+        nsec = nsec - NANOS_PER_SEC
+    end
+    return secs, nsec
+end
+
+local function datetime_cmp(lhs, rhs)
+    if not is_datetime(lhs) or not is_datetime(rhs) then
+        return nil
+    end
+    local sdiff = lhs.epoch - rhs.epoch
+    return sdiff ~= 0 and sdiff or (lhs.nsec - rhs.nsec)
+end
+
+local function datetime_eq(lhs, rhs)
+    local rc = datetime_cmp(lhs, rhs)
+    return rc ~= nil and rc == 0
+end
+
+local function datetime_lt(lhs, rhs)
+    local rc = datetime_cmp(lhs, rhs)
+    return rc == nil and error('incompatible types for comparison', 2) or
+           rc < 0
+end
+
+local function datetime_le(lhs, rhs)
+    local rc = datetime_cmp(lhs, rhs)
+    return rc == nil and error('incompatible types for comparison', 2) or
+           rc <= 0
+end
+
+local function datetime_serialize(self)
+    return { epoch = self.epoch, nsec = self.nsec,
+             tzoffset = self.tzoffset, tzindex = 0 }
+end
+
+local parse_zone
+
+local function datetime_new_raw(epoch, nsec, tzoffset, tzindex)
+    local dt_obj = ffi.new(datetime_t)
+    dt_obj.epoch = epoch
+    dt_obj.nsec = nsec
+    dt_obj.tzoffset = tzoffset or 0
+    dt_obj.tzindex = tzindex or 0
+    return dt_obj
+end
+
+local function datetime_new_dt(dt, secs, fraction, offset)
+    local epochV = dt ~= nil and (builtin.tnt_dt_rdn(dt) - DT_EPOCH_1970_OFFSET) *
+                   SECS_PER_DAY or 0
+    local secsV = secs or 0
+    local fracV = fraction or 0
+    local ofsV = offset or 0
+    return datetime_new_raw(epochV + secsV - ofsV * 60, fracV, ofsV)
+end
+
+local function get_timezone(offset)
+    if type(offset) == 'number' then
+        return offset
+    elseif type(offset) == 'string' then
+        return parse_zone(offset)
+    end
+end
+
+-- create datetime given attribute values from obj
+local function datetime_new(obj)
+    if obj == nil or type(obj) ~= 'table' then
+        return datetime_new_raw(0, 0, 0)
+    end
+
+    local ymd = false
+    local hms = false
+    local dt = DT_EPOCH_1970_OFFSET -- default is 1970-01-01
+
+    local y = obj.year
+    if y ~= nil then
+        check_range(y, {1, 9999}, 'year')
+        ymd = true
+    end
+    local M = obj.month
+    if M ~= nil then
+        check_range(M, {1, 12}, 'month')
+        ymd = true
+    end
+    local d = obj.day
+    if d ~= nil then
+        check_range(d, {1, 31, -1}, 'day')
+        ymd = true
+    end
+    local h = obj.hour
+    if h ~= nil then
+        check_range(h, {0, 23}, 'hour')
+        hms = true
+    end
+    local m = obj.min
+    if m ~= nil then
+        check_range(m, {0, 59}, 'min')
+        hms = true
+    end
+    local s = obj.sec
+    if s ~= nil then
+        check_range(s, {0, 60}, 'sec')
+        hms = true
+    end
+    local nsec, usec, msec = obj.nsec, obj.usec, obj.msec
+    -- if there are separate nsec, usec, or msec provided then
+    -- timestamp should be integer
+    local int_ts = nsec ~= nil or usec ~= nil or msec ~= nil
```

Comment:
It is only used under a condition on the line 298. Please, move it into the condition's body.

Reply:
Thanks, for pointing my attention to this line of code, I've realized that there is no proper handling implemented (copied from `:set` :)) for `msec`, 'usec`, and 'nsec`. Here is the delta introduced after some refactoring:

```diff
@@ -263,11 +263,18 @@ local function get_timezone(offset)
     end
 end
 
+local function bool2int(b)
+    return b and 1 or 0
+end
+
 -- create datetime given attribute values from obj
 local function datetime_new(obj)
@@ -303,17 +310,29 @@ local function datetime_new(obj)
         check_range(s, {0, 60}, 'sec')
         hms = true
     end
-    local nsec, usec, msec = obj.nsec, obj.usec, obj.msec
-    -- if there are separate nsec, usec, or msec provided then
-    -- timestamp should be integer
-    local int_ts = nsec ~= nil or usec ~= nil or msec ~= nil
 
+    local nsec, usec, msec = obj.nsec, obj.usec, obj.msec
+    local count_usec = bool2int(nsec ~= nil) + bool2int(usec ~= nil) +
+                       bool2int(msec ~= nil)
+    if count_usec > 1 then
+        error('only one of nsec, usec or msecs may defined simultaneously', 2)
+    end
+    if usec ~= nil then
+        nsec = usec * 1e3
+    elseif msec ~= nil then
+        nsec = msec * 1e6
+    end
     local ts = obj.timestamp
-    local fraction
     if ts ~= nil then
+        local fraction
         s, fraction = math_modf(ts)
-        if not int_ts then
+        -- if there are separate nsec, usec, or msec provided then
+        -- timestamp should be integer
+        if count_usec == 0 then
             nsec = fraction * 1e9
+        elseif fraction ~= 0 then
+            error('only integer values allowed in timestamp '..
+                  'if nsec, usec, or msecs provided', 2)
         end
         hms = true
     end
@@ -332,7 +351,8 @@ local function datetime_new(obj)
     if ymd then
         y = y or 1970
         M = M or 1
-        if d ~= nil and d < 0 then
+        d = d or 1
+        if d < 0 then
             d = builtin.tnt_dt_days_in_month(y, M)
         end
         dt = builtin.tnt_dt_from_ymd(y, M, d)
@@ -582,10 +602,6 @@ local function datetime_hms_update(self, h, m, s, new_offset)
     self.epoch = utc_secs(secs_day + h * 3600 + m * 60 + s, new_offset)
 end
 
-local function bool2int(b)
-    return b and 1 or 0
-end
-
 local function datetime_set(self, obj)
     check_table(obj, "datetime.set()")
 
```

Also I've covered all known attributes for passing to `date.new{}`, so current coverage is:
```
File                 Hits Missed Coverage
-----------------------------------------
builtin/datetime.lua 309  42     88.03%
-----------------------------------------
Total                309  42     88.03%
```

Reply:
Why isn't the coverage 100%? What couldn't you test?

Reply:
Yea, that's annoying me as well. But [luacov uses standard Lua/LuaJIT debugging/tracing facilities](https://github.com/keplerproject/luacov/blob/master/src/luacov/runner.lua#L452) and (apparently) counting original lines in module (I may be getting that's wrong). So if there are some multi-line `ffi.cdef [[ ]]` blocks - you could never count their lines executed. And by some odd reason, probably due to loading order, it never counts initialization expressions of upvalues.
Eventually I'll investigate it further.

Reply:
Yesterday, when I've achieved 93.77% of code covered, I've reevaluated these misses:

```
==============================================================================
Summary
==============================================================================
File                 Hits Missed Coverage
-----------------------------------------
builtin/datetime.lua 346  23     93.77%
-----------------------------------------
Total                346  23     93.77%
```
Let see where these `***0` were located:
```
✔ ~/datetime/tarantoolt/build [tsafin/gh-5941-datetime-take2-wip ↑·3|✚ 5…7⚑ 3]
22:16 $ fgrep '****0' luacov.report.out
****0 local ffi = require('ffi')
****0 ffi.cdef [[
****0 ]]
****0 local builtin = ffi.C
****0 local math_modf = math.modf
****0 local math_floor = math.floor
****0 local SECS_PER_DAY     = 86400
****0 local NANOS_PER_SEC    = 1000000000
****0 local DAYS_EPOCH_OFFSET = 719163
****0 local SECS_EPOCH_OFFSET = DAYS_EPOCH_OFFSET * SECS_PER_DAY
****0 local TOSTRING_BUFSIZE = 48
****0 local datetime_t = ffi.typeof('struct datetime')
****0         return error(("%s: expected datetime, but received %s"):
****0         return error(("%s: expected string, but received %s"):
****0         nsec = nsec % NANOS_PER_SEC
****0         return nil
****0     return { epoch = self.epoch, nsec = self.nsec,
****0              tzoffset = self.tzoffset, tzindex = 0 }
****0         secs = secs + direction * 60 * 60 * hours
****0         secs = secs + direction * 60 * minutes
****0         local s, frac = math_modf(seconds)
****0 ffi.metatype(datetime_t, {
****0 return setmetatable(
```

i.e. a number of local variable defines and meta operations, then a few non-visited lines:

```
✔ ~/datetime/tarantoolt/build [tsafin/gh-5941-datetime-take2-wip ↑·3|✚ 5…7⚑ 3]
22:21 $ fgrep '****0' luacov.report.out | egrep 'local|ffi|setmeta' | wc -l
14
```

i.e. out of 23 missed lines, 14 are non-fixable, and were uncounted due to late `luacov` module load (or due to limitations of approach used by `luacov` for tracing of executed/parsed lines). 9 - are theoretically fixable, but most of them at the moment, due to uncovered corners of interval support function (`datetime_increment`) and I'll rather fix it at the next step, not now, once I'll return proper interval operations from older branches.